src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -258,24 +239,266 @@ private IISAwaitable DoFlushAsync()
+                    else if (read != -1)
+                    {
+                        Input.Writer.Advance(read);
+                    }
+                }
+                finally
+                {
+                    // Always commit any changes to the Input pipe
+                    Input.Writer.Commit();
+                    _inputHandle.Dispose();
+                }
+
+                // Flush the read data for the Input Pipe writer
+                var flushResult = await Input.Writer.FlushAsync();
+
+                // If the pipe was closed, we are done reading, 
+                if (flushResult.IsCompleted || flushResult.IsCancelled)
+                {
+                    _doneReading = true;
+                }
+
+                // We are done reading now, set _reading to false.
+                // _reading may already be set to false through cancellation, but it is okay to be redundant here.
+                lock(_stateSync)
+                {
+                    _reading = false;
+                }
+            }
+
+            // Complete the input writer as we are done reading the request body.
+            Input.Writer.Complete();
+        }
+
+        private async Task WriteLoopAsync()
+        {
+            // At this point, reading is done, so we will await all reads from the output pipe
+            while (true)
+            {
+                // Always await every write as we are done reading.
+                var readResult = await Output.Reader.ReadAsync();
+
+                // Get data from pipe
+                var buffer = readResult.Buffer;
+                var consumed = buffer.End;
+
+                try
+                {
+                    // If the output pipe is termniated, cancel the request.
+                    if (readResult.IsCancelled)
+                    {
+                        break;
+                    }
+
+                    if (!buffer.IsEmpty)
+                    {
+                        // Write to IIS buffers
+                        // Guaranteed to write the entire buffer to IIS
+                        await WriteToIISAsync(buffer);
+                    }
+                    else if (readResult.IsCompleted)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        // Flush of zero bytes will 
+                        await FlushToIISAsync();
+                    }
+                }
+                finally
+                {
+                    // Always Advance the data pointer to the end of the buffer.
+                    Output.Reader.AdvanceTo(buffer.End);
+                }
+                await CheckForUpgrade();
+            }
+
+            // Close the output pipe as we are done reading from it.
+            Output.Reader.Complete();
+        }
+
+        private unsafe IISAwaitable FlushToIISAsync()
+        {
+            // Calls flush 
+            lock(_stateSync)
```

Comment:
What state are you protecting here?

Reply:
You are correct. I think I only need to do it on ReadFromIISAsync